numpyro/contrib/control_flow/scan.py

```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
                                f" but got {value_ndim}. Please report the issue to us!")
 
 
-def scan_wrapper(f, init, xs, length, reverse, rng_key=None, substitute_stack=[]):
+class promote_shapes(Messenger):
+    # a helper messenger to promote shapes of `fn` and `value`
+    def process_message(self, msg):
+        if msg["type"] == "sample" and msg["value"] is not None:
+            fn, value = msg["fn"], msg["value"]
+            value_batch_ndims = jnp.ndim(value) - fn.event_dim
+            fn_batch_ndim = len(fn.batch_shape)
+            prepend_shapes = (1,) * abs(fn_batch_ndim - value_batch_ndims)
+            if fn_batch_ndim > value_batch_ndims:
+                msg["value"] = jnp.reshape(value, prepend_shapes + jnp.shape(value))
+            elif fn_batch_ndim < value_batch_ndims:
+                msg["fn"] = tree_map(lambda x: jnp.reshape(x, prepend_shapes + jnp.shape(x)), fn)
+
+
+def scan_enum(f, init, xs, length, reverse, rng_key=None, substitute_stack=None):
+    # XXX: This implementation only works for history size=1 but can be
+    # extended to history size > 1 by running `f` `history_size` times
+    # for initialization. However, `sequential_sum_product` does not
+    # support history size > 1, so we skip supporting it here.
+    if reverse:
+        x0 = tree_map(lambda x: x[-1], xs)
+        xs_ = tree_map(lambda x: x[:-1], xs)
+    else:
+        x0 = tree_map(lambda x: x[0], xs)
+        xs_ = tree_map(lambda x: x[1:], xs)
+
+    carry_shape_at_t1 = None
+
+    def body_fn(wrapped_carry, x, prefix=None):
+        i, rng_key, carry = wrapped_carry
+        init = True if (not_jax_tracer(i) and i == 0) else False
+        rng_key, subkey = random.split(rng_key) if rng_key is not None else (None, None)
+
+        seeded_fn = handlers.seed(f, subkey) if subkey is not None else f
+        for subs_type, subs_map in substitute_stack:
+            subs_fn = partial(_subs_wrapper, subs_map, i, length)
+            if subs_type == 'condition':
+                seeded_fn = handlers.condition(seeded_fn, condition_fn=subs_fn)
+            elif subs_type == 'substitute':
+                seeded_fn = handlers.substitute(seeded_fn, substitute_fn=subs_fn)
+
+        if init:
+            with handlers.scope(prefix="_init"):
+                new_carry, y = seeded_fn(carry, x)
+                trace = {}
+        else:
+            with handlers.block(), packed_trace() as trace, promote_shapes(), enum(), markov():
+                new_carry, y = config_enumerate(seeded_fn)(carry, x)
+
+            # store shape of new_carry at a global variable
+            nonlocal carry_shape_at_t1
+            carry_shape_at_t1 = [jnp.shape(x) for x in tree_flatten(new_carry)[0]]
+            # make new_carry have the same shape as carry
+            # FIXME: is this rigorous?
```

Comment:
You may want to add some unit tests for this

Reply:
The thing is it passes all the test and I can't figure out when it possibly fails. :(

Reply:
Things may get more complicated if you decide to support `history > 1` and there are enumerated variables with different periods.